Add class to allow running depletion simulation independent of the OpenMC transport solver (#2100)
openmc-dev/openmc: docs/source/usersguide/depletion.rst

Depletion and Transmutation
===========================

OpenMC supports transport-coupled and transport-independent depletion, or
burnup, calculations through the :mod:`openmc.deplete` Python module. OpenMC
uses transmutation reaction rates to solve a set of transmutation equations
that determine the evolution of nuclide densities within a material. The

> nuclide densities predicted as some future time are then used to determine

Reviewer: ```suggestion
nuclide densities predicted at some future time are then used to determine
```

:class:`openmc.deplete.Operator` (which uses the OpenMC transport solver), but
in principle additional operator classes based on other transport codes could be
implemented and no changes to the depletion solver itself would be needed. The

> operator class requires a :class:`openmc.model.Model` instance containing 

Reviewer: ```suggestion
operator class requires a :class:`~openmc.Model` instance containing 
```

material, geometry, and settings information::

>     model = openmc.model.Model()

Reviewer: ```suggestion
    model = openmc.Model()
```

    results = openmc.deplete.Results("depletion_results.h5")
    time, keff = results.get_keff()

Note that the coupling between the reaction rate solver and the transmutation

> solver happens in-memory rather than by reading/writing files on disk. OpenMC has two categories of transport operators for obtaining transmutation reaction

Reviewer: ```suggestion
solver happens in-memory rather than by reading/writing files on disk. OpenMC
has two categories of transport operators for obtaining transmutation reaction
```

    # use a dictionary of fission_q values
    fission_q = {"U235": 202e+6}  # energy in eV

    # create a Model object

>     model  = openmc.model.Model(geometry, settings)

Reviewer: ```suggestion
    model  = openmc.Model(geometry, settings)
```

    This will increase the total memory usage and run time due to an increased
    number of tallies and material definitions.

Transport-independent depletion
-------------------------------

.. note::
   
   This is a brand-new feature and is under heavy development. API changes are
   possible and likely in the near future.

OpenMC also supports transport-independent depletion calculations using the
:class:`FluxDepletionOperator` class. Rather than taking a
:class:`openmc.model.Model` object, this class accepts a volume,
a dictionary of nuclide concentrations, a flux spectra, and one-group
microscopic cross sections as a pandas dataframe. The class includes
helper functions to constructe the dataframe from a csv file or from
data arrays::

    ...

>     micro_xs = FluxDepletionOperator.create_micro_xs_from_csv(micro_xs_path)

Author: I'm just trying to demonstrate how one would initiate a `FluxDepletionOperator` object. I have some discussion on the structure of the `micro_xs` parameter in the docstrings for that class, but I can try to add some details about that in the guide if you think it'd be useful.

    This will increase the total memory usage and run time due to an increased
    number of tallies and material definitions.

Transport-independent depletion

> -------------------------------

Reviewer: ```suggestion
===============================
```

.. note::
   
   This feature is still under heavy development. API changes are
   possible and likely in the near future.

OpenMC supports running depletion calculations independent of the OpenMC

> transport solver using the :class:`FluxDepletionOperator` class. This class

Reviewer: ```suggestion
transport solver using the :class:`~openmc.deplete.FluxDepletionOperator` class. This class
```

has two ways to initalize it; the default constructor accepts an 
:class:`openmc.Materials` object, a flux spectra, and one-group microscopic 

> cross sections as a :class:`pandas.Dataframe`, while the `from_nuclides` 

Reviewer: ```suggestion
cross sections as a :class:`pandas.DataFrame`, while the `from_nuclides` 
```

    This will increase the total memory usage and run time due to an increased
    number of tallies and material definitions.

Transport-independent depletion
===============================

.. note::
   
   This feature is still under heavy development. API changes are
   possible and likely in the near future.

OpenMC supports running depletion calculations independent of the OpenMC
transport solver using the :class:`~openmc.deplete.FluxDepletionOperator` class.
This class supports both constant-flux (``source-rate`` normalization) and 
constant-power depletion (``fission-q`` normalization).

.. important::

   Make sure you set the correct parameter in the :class:`openmc.abc.Integrator` class. Use the ``source_rates`` parameter when ``normalization_mode == source-rate``, and use ``power`` or ``power_density`` when ``normalization_mode == fission-q``.

.. warning::

>    The accuracy of results when using ``fission-q`` is entirely dependent on your depletion chain. Make sure it has sufficient data to resolve the dynamics of your particular scenario. 

Author: Sure thing

    This will increase the total memory usage and run time due to an increased
    number of tallies and material definitions.

Transport-independent depletion
===============================

> .. note::

Reviewer: Maybe change this to warning?
```suggestion
.. warning::
```

   
   This feature is still under heavy development and has yet to be rigorously 
   verified. API changes and feature additions are possible and likely in
   the near future.

This category of operator uses pre-calculated one-group microscopic cross
sections to obtain transmutation reaction rates. OpenMC provides the
:class:`~openmc.deplete.IndependentOperator` for this method of calculation.
While the one-group microscopic cross sections can be calculated using a
transport solver, :class:`~openmc.deplete.IndependentOperator` is not directly
coupled to any transport solver. The
:class:`~openmc.deplete.IndependentOperator` class requires a
:class:`openmc.Materials` object, a :class:`~openmc.deplete.MicroXS` object,
and a path to a depletion chain file::

    # load in the microscopic cross sections
    materials = openmc.Materials()
    ...

    micro_xs = openmc.deplete.MicroXS()
    ...

>     op = IndependentOperator(materials, micro_xs, chain_file)

Reviewer: ```suggestion
    op = openmc.deplete.IndependentOperator(materials, micro_xs, chain_file)
```

                                                          chain_file,
                                                          nuc_units='atom/cm3')

A user can then define an integrator class as they would for a coupled
transport-depletion calculation and follow the same steps from there.

.. note::

   Ideally, one-group cross section data should be available for every
   reaction in the depletion chain. If a nuclide that has a reaction 
   associated with it in the depletion chain is present in the `nuclides` 
   parameter but not the cross section data, that reaction will not be
   simulated.

Generating Microscopic Cross Sections
-------------------------------------

Users can generate the one-group microscopic cross sections needed by
:class:`~openmc.deplete.IndependentOperator` using the
:class:`~openmc.deplete.MicroXS` class::

    import openmc

    model = openmc.Model.from_xml()

    micro_xs = openmc.deplete.MicroXS.from_model(model, model.materials[0])

The :meth:`~openmc.deplete.MicroXS.from_model()` method will produce a
:class:`~openmc.deplete.MicroXS` object with microscopic cross section data in

> units of ``b``, which is what :class:`~openmc.deplete.IndependentOperator`

Reviewer: ```suggestion
units of barns, which is what :class:`~openmc.deplete.IndependentOperator`
```

> expects the units to be. The :class:`~openmc.deplete.MicroXS` class also includes functions to read in cross section data directly from a ``.csv`` file or from data arrays::

Reviewer: ```suggestion
expects the units to be. The :class:`~openmc.deplete.MicroXS` class also
includes functions to read in cross section data directly from a ``.csv`` file
or from data arrays::
```

    This will increase the total memory usage and run time due to an increased
    number of tallies and material definitions.

Transport-independent depletion
===============================

.. note::
   

>    This feature is still under heavy development and has yet to be verifed
>    code-to-code . API changes and feature additions are possible and likely in
>    the near future.

Reviewer: ```suggestion
   This feature is still under heavy development and has yet to be rigorously
   verified. API changes and feature additions are possible and likely in
   the near future.
```

   parameter but not the cross section data, that reaction will not be
   simulated.

Generating Microscopic Cross Sections
-------------------------------------

Users can generate the one-group microscopic cross sections needed by
:class:`~openmc.deplete.IndependentOperator` using the
:class:`~openmc.deplete.MicroXS` class::

    import openmc

    model = openmc.Model.from_xml()

    micro_xs = openmc.deplete.MicroXS.from_model(model, model.materials[0])

The :meth:`~openmc.deplete.MicroXS.from_model()` method will produce a
:class:`~openmc.deplete.MicroXS` object with microscopic cross section data in
units of ``b``, which is what :class:`~openmc.deplete.IndependentOperator`
expects the units to be. The :class:`~openmc.deplete.MicroXS` class also includes functions to read in cross section data directly from a ``.csv`` file or from data arrays::

    micro_xs = MicroXS.from_csv(micro_xs_path)

    nuclides = ['U234', 'U235', 'U238']
    reactions = ['fission', '(n,gamma)']
    data = np.array([[0.1, 0.2],
                     [0.3, 0.4],
                     [0.01, 0.5]])
    micro_xs = MicroXS.from_array(nuclides, reactions, data)

> .. important ::

Reviewer: ```suggestion
.. important::
```

   Both :meth:`~openmc.deplete.MicroXS.from_csv()` and
   :meth:`~openmc.deplete.MicroXS.from_array()` assume the cross section values 
   provided are in barns by defualt, but have no way of verifying this. Make
   sure your cross sections are in the correct units before passing to a
   :class:`~openmc.deplete.IndependentOperator` object.

Caveats
-------

Reaction Rate Normalization
~~~~~~~~~~~~~~~~~~~~~~~~~~~

The :class:`~openmc.deplete.IndependentOperator` class supports two methods for
normalizing reaction rates:

.. important::

   Make sure you set the correct parameter in the :class:`openmc.abc.Integrator`
   class. Use the ``source_rates`` parameter when
   ``normalization_mode == source-rate``, and use ``power`` or ``power_density``
   when ``normalization_mode == fission-q``.

> 1. ``soure-rate`` normalization, which assumes the ``source-rate`` provided by

Reviewer: ```suggestion
1. ``source-rate`` normalization, which assumes the ``source-rate`` provided by
```

   the time integrator is a flux, and obtains the reaction rates by multiplying
   the cross-sections by the ``source-rate``.

> 2. ``fission-q`` normalization, which assumes the ``source-rate`` provided by

Reviewer: In this case, doesn't the user pass `power` or `power_density`, not `source-rate`?

Author: Correct. I think I was referring to how internally in `Integrator`, the variable passed to the operator classes no matter what the user provides is `source_rate`, but that doesn't really make a difference to the user. I'll adjust the wording here.

   the time integrator is a power, and obtains the reaction rates by computing a
   value for the flux based on this power. The general equation for the flux is 

   .. math::

      \phi = \frac{P}{V \cdot \sum_i (Q_i \cdot \Sigma^f_i \cdot \rho_i)}

   where :math:`\sum_i` is the sum over all nuclides :math:`i`. This equation
   makes the same assumptions and issues as discussed in
   :ref:`energy-deposition`. Unfortunately, the proposed solution in that
   section does not apply here since we are decoupled from transport code.
   However, there is a method to converge to a more accurate value for flux by
   using substeps during time integration.
   `This paper <https://doi.org/10.1016/j.anucene.2016.05.031>`_ provides a

>    good discussion of this method. Hopefully such a method will be implemented
>    in OpenMC in the near future.

Reviewer: I try to avoid making promises in the documentation
```suggestion
   good discussion of this method.
```

Author: haha that's a good recommendation

.. warning::

   The accuracy of results when using ``fission-q`` is entirely dependent on
   your depletion chain. Make sure it has sufficient data to resolve the
   dynamics of your particular scenario. 

Multiple Materials
~~~~~~~~~~~~~~~~~~

Running a depletion simulation with multiple materials using the
``source-rate`` normalization method treats each material as completely
separate with respect to reaction rates. This can be useful for running many
different cases of a particular scenario. However, running a depletion
simulation with multiple materials using the ``fission-q`` normalization method
treats each material as part of the same "reactor" due to how ``fission-q``

> normalization conglomerates energy values from each material to a single value.

Reviewer: ```suggestion
normalization accumulates energy values from each material to a single value.
```

    This will increase the total memory usage and run time due to an increased
    number of tallies and material definitions.

Transport-independent depletion
===============================

.. warning::
   
   This feature is still under heavy development and has yet to be rigorously 
   verified. API changes and feature additions are possible and likely in
   the near future.

This category of operator uses pre-calculated one-group microscopic cross
sections to obtain transmutation reaction rates. OpenMC provides the
:class:`~openmc.deplete.IndependentOperator` for this method of calculation.
While the one-group microscopic cross sections can be calculated using a
transport solver, :class:`~openmc.deplete.IndependentOperator` is not directly
coupled to any transport solver. The
:class:`~openmc.deplete.IndependentOperator` class requires a
:class:`openmc.Materials` object, a :class:`~openmc.deplete.MicroXS` object,
and a path to a depletion chain file::

    # load in the microscopic cross sections
    materials = openmc.Materials()
    ...

>     micro_xs = openmc.deplete.MicroXS

Reviewer: Probably should add `from_...(...)`?
